common/djangoapps/student/models.py

```diff
@@ -1535,7 +1535,7 @@ def generate_enrollment_status_hash(cls, user):
 
         if not status_hash:
             enrollments = cls.enrollments_for_user(user).values_list('course_id', 'mode')
-            enrollments = [(e[0].lower(), e[1].lower()) for e in enrollments]
+            enrollments = [(unicode(e[0]).lower(), e[1].lower()) for e in enrollments]
```

Comment:
The `unicode` in this case simply converts a CourseKey object to a string-representation of the CourseKey. It doesn't do any conversion of the actual characters in the course key.

Reply:
Why convert to unicode at all?

Reply:
FWIW, in python2 the CourseLocator object does supply a `__str__` function which would properly return encoded unicode byte sequences if the course id actually does contain non-ascii.

Say we change this to str().  As is, things stay the same because we don't have unicode course ids (crossing fingers).  Then, if in the future we decide to officially support unicode in course ids, the format() wouldn't raise an exception, and the hasher is fed raw unicode byte sequences.  So far so good.  If we switch over to python3, the hash input still shouldn't change since it all gets flattened into bytestrings anyway due to the `.encode('utf-8')` below.

Reply:
The `unicode` here only serves to serialize a CourseKey object to its string representation:
https://github.com/edx/edx-platform/wiki/Opaque-Keys-(Locators)#serialize-keys-to-strings
I'm willing to change this to `str` instead - but that's *not* the standard we've used throughout the code base. In Python3, there's admittedly no `unicode()` method. But I defaulted to sticking with the standard for now.

Reply:
If it is similar to the rest of the codebase, then I agree about leaving it unicode().  In retrospect, I'm not really sure how lower() behaves when faced with unicode bytes vs. unicode characters.